### ISSUE TITLE
Nodemon process crashing when trying to re-attach debugger

### ISSUE THREAD

**Issue by auchenberg** [Microsoft/vscode-node-debug2#154]
I'm in the middle of writing a NodeMon recipe and found an interesting issue when trying to configure our Node debugger in a launch config with the `restart` option.

Node version: v8.9.1

### Steps:
1. Checkout https://github.com/gtsopour/nodejs-shopping-cart
2. Update `package.json` to have a Nodemon script like:
```
    "scripts": {
        "start": "node ./bin/www",
        "nodemon": "nodemon --inspect ./bin/www"
    },
```
3.  Use the following launch config:
```
        {
            "type": "node",
            "request": "launch",
            "name": "Node: Nodemon",
            "runtimeExecutable": "npm",
            "runtimeArgs": ["run-script", "nodemon"],
            "port": 9229,
            "restart": true,
            "trace": "verbose"
        }
```
4. Start the debug config
5. Debugger gets attached
6. Set breakpoint in `index.js line: 10`
7. open http://localhost:3000
8. Breakpoints get's hit
9. Make change to `line 10` ex update: `res.render('index2'`
10. Nodemon kicks in, restartes the process
11. Exception get's thrown
```
From target: {"method":"Runtime.exceptionThrown","params":{"timestamp":4227945.763027,"exceptionDetails":{"exceptionId":1,"text":"Uncaught","lineNumber":192,"columnNumber":5,"scriptId":"17","url":"events.js","stackTrace":{"callFrames":[{"functionName":"_errnoException","scriptId":"18","url":"util.js","lineNumber":1030,"columnNumber":12},{"functionName":"afterWrite","scriptId":"64","url":"net.js","lineNumber":872,"columnNumber":13}],"parent":{"description":"WRITEWRAP","callFrames":[{"functionName":"init","scriptId":"34","url":"internal/inspector_async_hook.js","lineNumber":18,"columnNumber":14},{"functionName":"emitInitNative","scriptId":"29","url":"async_hooks.js","lineNumber":459,"columnNumber":42},{"functionName":"createWriteReq","scriptId":"64","url":"net.js","lineNumber":805,"columnNumber":20},{"functionName":"Socket._writeGeneric","scriptId":"64","url":"net.js","lineNumber":768,"columnNumber":10},{"functionName":"Socket._write","scriptId":"64","url":"net.js","lineNumber":791,"columnNumber":7},{"functionName":"doWrite","scriptId":"50","url":"_stream_writable.js","lineNumber":387,"columnNumber":11},{"functionName":"writeOrBuffer","scriptId":"50","url":"_stream_writable.js","lineNumber":373,"columnNumber":4},{"functionName":"Writable.write","scriptId":"50","url":"_stream_writable.js","lineNumber":290,"columnNumber":10},{"functionName":"Socket.write","scriptId":"64","url":"net.js","lineNumber":709,"columnNumber":39},{"functionName":"logRequest","scriptId":"164","url":"/Use[...]
To client: {"seq":0,"type":"event","event":"output","body":{"category":"stderr","output":"Error: write EPIPE\n    at _errnoException (util.js:1031:13)\n    at WriteWrap.afterWrite [as oncomplete] (net.js:873:14)\n","source":{"name":"util.js","path":"util.js","sourceReference":1002,"origin":"read-only core module"},"line":1031,"column":13}}
Error: write EPIPE
util.js:1031
    at _errnoException (util.js:1031:13)
    at WriteWrap.afterWrite [as oncomplete] (net.js:873:14)
← From target: {"method":"Runtime.executionContextDestroyed","params":{"executionContextId":1}}
Killing process with id: 2552
Terminated: Program ended
```
12. Notice that we'll then `Killing process with id: 2552` the process, which causes failure in re-attaching.

### Attach config works
If run my `npm run nodemon` script from a terminal and uses this attach config, then our debugger can attach and re-attach for the same changes. I don't see an exception being thrown in the logs.
```
        {
            "type": "node",
            "request": "attach",
            "name": "Attach by Process ID",
            "processId": "${command:PickProcess}",
            "restart": true,
            "protocol": "inspector",
            "trace": "verbose"
        }
```

### Thoughts
So it seems to be something related to the way we spawn the NPM script that's causing problems. I'm wondering what that could be as the exception seem to come from some node internals.

**Comment by roblourens** [comment]
I think my fix for Microsoft/vscode#35739 exposed the adapter killing the process at the wrong time. If so a workaround would be `"console": "integratedTerminal"`

**Comment by roblourens** [comment]
Quick brain dump because I'm sure I'll be back here in the future...

Originally, the adapter would try to get the PID from the process it spawns, then also get the PID from the process it attaches to. Then when the connection is closed for any reason, it kills the process with that PID.

The problem in Microsoft/vscode#35739 is that we attach to a Node process inside Docker which has a PID of 1 in Docker's own PID space. Then we'd try to kill PID 1 and take down the whole machine. The fix was to not replace the PID that we spawned with the PID that we attached to. Without trying it, I think it probably still fails if you launch Docker in the integrated terminal.

So the problem here is that previously we had the PID of the attached node process, but now when launching in the console we just have the pid of nodemon itself. And when the node process restarts, we kill nodemon. The fix is to, when the restart flag is set, not kill anything unless we first get a 'disconnect' request from the client. That ensures that we kill nodemon when the user clicks the stop button, but not when the node process restarts.

I'm not 100% sure this is right. It raises an interesting issue re: killing the spawned process vs killing the attached process. I think killing the spawned process is a safer bet for cleaning everything up. But it's possible that this isn't the right thing to do for some tool.

/cc @weinand
